subprojects/zydis-4.1.1/man/ZydisDisasm.1.ronn
ZydisDisasm(1) -- disassemble files
===================================

## SYNOPSIS

`ZydisDisasm` <machine_mode> [<input_file>]

## DESCRIPTION

`ZydisDisasm` allows you to decode X86 & X86-64 assembly files, dumping the disassembled instructions to stdout. With no <input_file> argument, `ZydisDisasm` will read input from stdin.

## OPTIONS

`ZydisDisasm` supports four different machine modes

  * `-real`:
    real machine mode

  * `-16`:
    16 bits machine mode

  * `-32`:
    32 bits machine mode

  * `-64`:
    64 bits machine mode

## EXAMPLES

    $ ZydisDisasm -64 input.hex
    and byte ptr ds:[rbx], dh
    and byte ptr ds:[r14], r14b
    xor eax, 0x20453220
    xor byte ptr ds:[rax], r12b
    xor r12b, byte ptr ds:[rax]
    xor r12d, dword ptr ds:[rax]
    xor al, 0x38
    and byte ptr ds:[rax], dh
    xor dword ptr ds:[rax], esp
    xor al, 0x20
    cmp dword ptr ds:[rax], edi
    and byte ptr ds:[rdx], dh
    and byte ptr ds:[r8], sil
    xor dword ptr ds:[rax], esp
    xor byte ptr ds:[rax], dh
    and byte ptr ds:[rax], dh
    xor byte ptr ds:[rdx], cl

## SEE ALSO

ZydisInfo(1)


ChangeLog
-=== RELEASE 0.8 ===-
changeset:   1625:0aaad74ad9fc
bookmark:    master
tag:         tip
user:        pancake
date:        Fri Jul 15 18:13:00 2011 +0200
files:       configure configure.acr libr/core/core.c libr/io/undo.c r2-bindings/python/r_cons.i.experimental
description:
* Bump release 0.8 - Codename thecakeisapie
* Fix uninitialized memory issue in r_block_resize()
  - Thanks vext01 for reporting


changeset:   1624:1315774824fb
parent:      1622:fc2643c36e30
parent:      1623:95f09764122e
user:        pancake
date:        Fri Jul 15 18:07:56 2011 +0200
description:
* Merge heads


changeset:   1623:95f09764122e
parent:      1620:236b81674ab5
user:        earada
date:        Fri Jul 15 14:17:36 2011 +0200
files:       libr/bin/p/bin_elf.c
description:
* Fix virtual address in elf without setcions


changeset:   1622:fc2643c36e30
user:        pancake
date:        Fri Jul 15 12:57:30 2011 +0200
files:       mk/gitpush.sh
description:
* Fix some typos in gitpush script


changeset:   1621:b6a6d3012998
user:        pancake
date:        Fri Jul 15 12:48:46 2011 +0200
files:       Makefile mk/gitpush.sh
description:
* Added gitpush utility for github mirroring


changeset:   1620:236b81674ab5
user:        pancake@pair
date:        Fri Jul 15 02:21:40 2011 +0200
files:       configure configure.acr
description:
* Bump acr to 0.8.6 with shellscript endian check
  - New flags --with-{little|big}-endian


changeset:   1619:d49dd6f2e789
user:        pancake@pair
date:        Fri Jul 15 01:50:42 2011 +0200
files:       binr/Makefile doc/maemo libr/Makefile maemo/Makefile maemo/README maemo/deb_hand.mak maemo/description.txt
description:
* Added support for new platform Maemo6-Harmattan
  - Linux/arm with debugger support
  - Debian package generation tools (no need for dpkg)
  - Compilation instructions at doc/maemo
* Fix install in binr and libr


changeset:   1618:dd1c50f121bf
user:        pancake@pair
date:        Thu Jul 14 22:14:11 2011 +0200
files:       configure configure.acr
description:
* Handle -lssl and pkg-config openssl
  - Should fix NetBSD build


changeset:   1617:4c6e8c50e7cf
user:        earada
date:        Thu Jul 14 16:17:39 2011 +0200
files:       libr/debug/p/debug_native.c libr/util/sys.c
description:
* Fix FreeBSD build


changeset:   1616:fd36edc76e42
user:        pancake
date:        Wed Jul 13 19:37:28 2011 +0200
files:       test/t/asm
description:
* Fix asm test


changeset:   1615:1ad69f14f1db
user:        pancake
date:        Wed Jul 13 19:28:53 2011 +0200
files:       test/tests.sh
description:
* Fix test suite


changeset:   1614:6c552be332cb
user:        pancake
date:        Wed Jul 13 19:22:38 2011 +0200
files:       binr/radare2/Makefile binr/radiff2/Makefile binr/rafind2/Makefile config-user.mk.acr configure configure.acr libr/debug/t/Makefile libr/hash/t/Makefile libr/include/r_userconf.h.acr libr/io/p/debug.mk libr/io/p/gdb.mk libr/io/p/haret.mk libr/io/t/Makefile libr/socket/Makefile libr/socket/t/Makefile
description:
* Use OpenSSL (cflags/ldflags) from pkg-config


changeset:   1613:8ecc34f29bc4
user:        pancake
date:        Wed Jul 13 18:55:16 2011 +0200
files:       doc/windows libr/asm/arch/msil/demsil.c libr/asm/p/asm_dalvik.c libr/bin/format/mach0/dyldcache.c libr/config.mk.tail libr/core/vmenus.c libr/include/list.h libr/include/r_types_base.h libr/io/p/io_w32dbg.c libr/socket/t/test.c libr/util/cache.c
description:
* Fix build and lot of warnings for w64


changeset:   1612:1552f58ae98c
user:        earada
date:        Wed Jul 13 17:41:26 2011 +0200
files:       binr/radare2/radare2.c libr/core/cmd.c libr/core/config.c libr/core/file.c libr/include/r_core.h
description:
* Use native api to calculate md5/sha1 hashes
- Fix bug 29


changeset:   1611:3dd0b00836fe
user:        pancake
date:        Wed Jul 13 13:59:55 2011 +0200
files:       binr/rax2/rax2.c doc/fortunes libr/fs/Makefile libr/fs/p/grub/Makefile libr/include/r_io.h libr/io/p/io_w32dbg.c
description:
* Fix w32 and w64 build (r_fs/grub issue)
* Dont quit rax2 when reading from stdin until 'q' or ^D
* Minor fix in fortunes


changeset:   1610:2637b35a1b31
user:        pancake
date:        Wed Jul 13 12:21:56 2011 +0200
files:       libr/core/cmd.c
description:
* Fix double pdf (thanks vext01)


changeset:   1609:5b507dbd2c5e
user:        pancake
date:        Wed Jul 13 11:30:25 2011 +0200
files:       Makefile
description:
* Remove manpages in purge
* Fix make symstall


changeset:   1608:980201d7e995
user:        pancake@pair
date:        Wed Jul 13 10:14:30 2011 +0200
files:       binr/binr.mk binr/radare2/Makefile binr/radiff2/Makefile binr/rafind2/Makefile doc/static libr/debug/p/libgdbwrap/gdbwrapper.c libr/debug/p/libgdbwrap/include/gdbwrapper-internals.h libr/debug/p/libgdbwrap/include/gdbwrapper.h libr/debug/t/Makefile libr/hash/t/Makefile libr/io/Makefile libr/io/p/debug.mk libr/io/p/gdb.mk libr/io/p/haret.mk libr/io/p/io_gdb.c libr/io/p/mach.mk libr/io/p/malloc.mk libr/io/p/shm.mk libr/io/t/Makefile libr/socket/t/Makefile mk/gcc.mk
description:
* Fix static build (--without-pic --with-nonpic)


changeset:   1607:3a06619b5226
user:        pancake@pair
date:        Tue Jul 12 23:22:33 2011 +0200
files:       libr/core/asm.c libr/core/cmd.c
description:
* Fix pi and /c commands


changeset:   1606:e220fccf6bc4
user:        earada
date:        Tue Jul 12 23:06:06 2011 +0200
files:       libr/core/cmd.c
description:
* Fix some warnings


changeset:   1605:d15cbb5291a3
user:        pancake
date:        Tue Jul 12 19:42:20 2011 +0200
files:       doc/windows libr/io/p/io_w32dbg.c mk/x86_64-w64-mingw32-gcc.mk
description:
* Fix build for ming-w64


changeset:   1604:026708c1cb88
user:        earada
date:        Tue Jul 12 17:26:56 2011 +0200
files:       libr/core/vmenus.c libr/print/format.c
description:
* Enhacements in keybindings used in visual mounts menus
* Implement missing 'pf p' pointer format


changeset:   1603:71f82c98175a
user:        pancake@pairch

subprojects/zydis-4.1.1/tools/ZydisDisasm.c
/***************************************************************************************************

  Zyan Disassembler Library (Zydis)

  Original Author : Florian Bernd, Joel Hoener

 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.

***************************************************************************************************/

/**
 * @file
 * Reads a byte-stream from a file or the `stdin` pipe and prints a textual
 * representation of the decoded data.
 */

#include "ZydisToolsShared.h"

#include <inttypes.h>
#include <stdio.h>

#include <Zycore/API/Terminal.h>
#include <Zycore/LibC.h>
#include <Zydis/Zydis.h>

#ifdef ZYAN_WINDOWS
#   include <fcntl.h>
#   include <io.h>
#endif

/* ============================================================================================== */
/* Colors                                                                                         */
/* ============================================================================================== */

/* ---------------------------------------------------------------------------------------------- */
/* Configuration                                                                                  */
/* ---------------------------------------------------------------------------------------------- */

#define COLOR_ADDRESS   ZYAN_VT100SGR_FG_BRIGHT_BLACK
#define COLOR_INVALID   ZYAN_VT100SGR_FG_BRIGHT_BLACK

/* ---------------------------------------------------------------------------------------------- */

/* ============================================================================================== */
/* Print functions                                                                                */
/* ============================================================================================== */

/**
 * Prints the instruction runtime address
 *
 * @param   runtime_address The runtime address of the instruction.
 */
static void PrintRuntimeAddress(ZyanU64 runtime_address)
{
    ZYAN_FPRINTF(ZYAN_STDOUT, "%s%016" PRIX64 "%s ",
        CVT100_ERR(COLOR_ADDRESS), runtime_address,
        CVT100_ERR(ZYAN_VT100SGR_RESET));
}

/**
 * Prints the formatted instruction disassembly.
 *
 * @param   formatter       A pointer to the `ZydisFormatter` instance.
 * @param   instruction     A pointer to the `ZydisDecodedInstruction` struct.
 * @param   operands        A pointer to the first `ZydisDecodedOperand` struct of the instruction.
 * @param   buffer          A pointer to the output buffer.
 * @param   length          The length of the output buffer (in bytes).
 * @param   runtime_address The runtime address of the instruction.
 */
static void PrintDisassembly(const ZydisFormatter* formatter,
    const ZydisDecodedInstruction* instruction, const ZydisDecodedOperand* operands,
    ZyanU8* buffer, ZyanUSize length, ZyanU64 runtime_address)
{
    ZyanStatus status;
    const ZydisFormatterToken* token;

    if (!ZYAN_SUCCESS(status = ZydisFormatterTokenizeInstruction(formatter, instruction, operands,
        instruction->operand_count_visible, buffer, length, runtime_address, &token, ZYAN_NULL)))
    {
        PrintStatusError(status, "Failed to tokenize instruction");
        exit(status);
    }

    PrintTokenizedInstruction(token);
}

/* ============================================================================================== */
/* Entry point                                                                                    */
/* ============================================================================================== */

void PrintUsage(int argc, char* argv[])
{
    ZYAN_FPRINTF(ZYAN_STDERR, "%sUsage: %s -[real|16|32|64] [input file]%s\n",
        CVT100_ERR(COLOR_ERROR), (argc > 0 ? argv[0] : "ZydisDisasm"),
        CVT100_ERR(ZYAN_VT100SGR_RESET));
}

int main(int argc, char** argv)
{
    InitVT100();

    if (ZydisGetVersion() != ZYDIS_VERSION)
    {
        ZYAN_FPRINTF(ZYAN_STDERR, "%sInvalid zydis version%s\n",
            CVT100_ERR(COLOR_ERROR), CVT100_ERR(ZYAN_VT100SGR_RESET));
        return EXIT_FAILURE;
    }

    if (argc < 2 || argc > 3)
    {
        PrintUsage(argc, argv);
        return EXIT_FAILURE;
    }

    ZydisDecoder decoder;
    if (!ZYAN_STRCMP(argv[1], "-real"))
    {
        ZydisDecoderInit(&decoder, ZYDIS_MACHINE_MODE_REAL_16, ZYDIS_STACK_WIDTH_16);
    }
    else if (!ZYAN_STRCMP(argv[1], "-16"))
    {
        ZydisDecoderInit(&decoder, ZYDIS_MACHINE_MODE_LONG_COMPAT_16, ZYDIS_STACK_WIDTH_16);
    }
    else if (!ZYAN_STRCMP(argv[1], "-32"))
    {
        ZydisDecoderInit(&decoder, ZYDIS_MACHINE_MODE_LONG_COMPAT_32, ZYDIS_STACK_WIDTH_32);
    }
    else if (!ZYAN_STRCMP(argv[1], "-64"))
    {
        ZydisDecoderInit(&decoder, ZYDIS_MACHINE_MODE_LONG_64, ZYDIS_STACK_WIDTH_64);
    }
    else
    {
        PrintUsage(argc, argv);
        return EXIT_FAILURE;
    }

    FILE* file = (argc >= 3) ? fopen(argv[2], "rb") : ZYAN_STDIN;
    if (!file)
    {
        ZYAN_FPRINTF(ZYAN_STDERR, "%sCan not open file '%s': %s%s\n",
            CVT100_ERR(COLOR_ERROR), argv[2], strerror(ZYAN_ERRNO),
            CVT100_ERR(ZYAN_VT100SGR_RESET));
        return EXIT_FAILURE;
    }
#ifdef ZYAN_WINDOWS
    // The `stdin` pipe uses text-mode on Windows platforms by default. We need it to be opened in
    // binary mode
    if (file == ZYAN_STDIN)
    {
        (void)_setmode(_fileno(ZYAN_STDIN), _O_BINARY);
    }
#endif

    ZydisFormatter formatter;
    if (!ZYAN_SUCCESS(ZydisFormatterInit(&formatter, ZYDIS_FORMATTER_STYLE_INTEL)) ||
        !ZYAN_SUCCESS(ZydisFormatterSetProperty(&formatter,
            ZYDIS_FORMATTER_PROP_FORCE_SEGMENT, ZYAN_TRUE)) ||
        !ZYAN_SUCCESS(ZydisFormatterSetProperty(&formatter,
            ZYDIS_FORMATTER_PROP_FORCE_SIZE, ZYAN_TRUE)))
    {
        ZYAN_FPRINTF(ZYAN_STDERR, "%sFailed to initialize instruction-formatter%s\n",
            CVT100_ERR(COLOR_ERROR), CVT100_ERR(ZYAN_VT100SGR_RESET));
        return EXIT_FAILURE;
    }

    ZyanU8 buffer[1024];
    ZyanUSize buffer_size;
    ZyanUSize buffer_remaining = 0;
    ZyanUSize read_offset_base = 0;
    do
    {
        buffer_size = fread(buffer + buffer_remaining, 1, sizeof(buffer) - buffer_remaining, file);
        if (buffer_size != (sizeof(buffer) - buffer_remaining))
        {
            if (ferror(file))
            {
                return EXIT_FAILURE;
            }
            ZYAN_ASSERT(feof(file));
        }
        buffer_size += buffer_remaining;

        ZydisDecodedInstruction instruction;
        ZydisDecodedOperand operands[ZYDIS_MAX_OPERAND_COUNT];
        ZyanStatus status;
        ZyanUSize read_offset = 0;
        ZyanU8 format_buffer[256];

        while ((status = ZydisDecoderDecodeFull(&decoder, buffer + read_offset,
            buffer_size - read_offset, &instruction, operands)) != ZYDIS_STATUS_NO_MORE_DATA)
        {
            const ZyanU64 runtime_address = read_offset_base + read_offset;

            PrintRuntimeAddress(runtime_address);

            if (!ZYAN_SUCCESS(status))
            {
                ZYAN_FPRINTF(ZYAN_STDOUT, "%sdb %02X%s\n", CVT100_OUT(COLOR_INVALID),
                    buffer[read_offset++], CVT100_OUT(ZYAN_VT100SGR_RESET));
                continue;
            }

            PrintDisassembly(&formatter, &instruction, operands, format_buffer,
                sizeof(format_buffer), runtime_address);

            read_offset += instruction.length;
        }

        buffer_remaining = 0;
        if (read_offset < sizeof(buffer))
        {
            buffer_remaining = sizeof(buffer) - read_offset;
            memmove(buffer, buffer + read_offset, buffer_remaining);
        }
        read_offset_base += read_offset;
    } while (buffer_size == sizeof(buffer));

    return EXIT_SUCCESS;
}

/* ============================================================================================== */
